File: taurus-protect-sdk-typescript/src/mappers/job.ts
/**
 * Job mapper functions for converting OpenAPI DTOs to domain models.
 */

import type { Job, JobStatistics, JobStatus } from '../models/job';
import { safeDate, safeMap, safeString } from './base';

/**
 * Maps a JobStatus DTO to a JobStatus domain model.
 */
export function jobStatusFromDto(dto: unknown): JobStatus | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    id: safeString(d.id),
    startedAt: safeDate(d.startedAt ?? d.started_at),
    updatedAt: safeDate(d.updatedAt ?? d.updated_at),
    timeoutAt: safeDate(d.timeoutAt ?? d.timeout_at),
    message: safeString(d.message),
    status: safeString(d.status),
  };
}

/**
 * Maps a JobStatistics DTO to a JobStatistics domain model.
 */
export function jobStatisticsFromDto(dto: unknown): JobStatistics | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    pending: safeString(d.pending),
    successes: safeString(d.successes),
    failures: safeString(d.failures),
    lastSuccess: jobStatusFromDto(d.lastSuccess ?? d.last_success),
    lastFailure: jobStatusFromDto(d.lastFailure ?? d.last_failure),
    avgDuration: safeString(d.avgDuration ?? d.avg_duration),
    maxDuration: safeString(d.maxDuration ?? d.max_duration),
    minDuration: safeString(d.minDuration ?? d.min_duration),
  };
}

/**
 * Maps a Job DTO to a Job domain model.
 */
export function jobFromDto(dto: unknown): Job | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    name: safeString(d.name),
    statistics: jobStatisticsFromDto(d.statistics),
  };
}

/**
 * Maps an array of Job DTOs to Job domain models.
 */
export function jobsFromDto(dtos: unknown[] | null | undefined): Job[] {
  return safeMap(dtos, jobFromDto);
}

File: taurus-protect-sdk-go/pkg/protect/mapper/job.go
package mapper

import (
	"strconv"

	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// JobFromDTO converts an OpenAPI TgvalidatordJob to a domain Job.
func JobFromDTO(dto *openapi.TgvalidatordJob) *model.Job {
	if dto == nil {
		return nil
	}

	job := &model.Job{
		Name: safeString(dto.Name),
	}

	if dto.Statistics != nil {
		job.Statistics = JobStatisticsFromDTO(dto.Statistics)
	}

	return job
}

// JobsFromDTO converts a slice of OpenAPI TgvalidatordJob to domain Jobs.
func JobsFromDTO(dtos []openapi.TgvalidatordJob) []*model.Job {
	if dtos == nil {
		return nil
	}
	jobs := make([]*model.Job, len(dtos))
	for i := range dtos {
		jobs[i] = JobFromDTO(&dtos[i])
	}
	return jobs
}

// JobStatisticsFromDTO converts an OpenAPI TgvalidatordJobStatistics to a domain JobStatistics.
func JobStatisticsFromDTO(dto *openapi.TgvalidatordJobStatistics) *model.JobStatistics {
	if dto == nil {
		return nil
	}

	stats := &model.JobStatistics{
		AvgDuration: safeString(dto.AvgDuration),
		MaxDuration: safeString(dto.MaxDuration),
		MinDuration: safeString(dto.MinDuration),
	}

	// Parse numeric string fields
	if dto.Pending != nil {
		if pending, err := strconv.ParseInt(*dto.Pending, 10, 64); err == nil {
			stats.Pending = pending
		}
	}
	if dto.Successes != nil {
		if successes, err := strconv.ParseInt(*dto.Successes, 10, 64); err == nil {
			stats.Successes = successes
		}
	}
	if dto.Failures != nil {
		if failures, err := strconv.ParseInt(*dto.Failures, 10, 64); err == nil {
			stats.Failures = failures
		}
	}

	// Convert job status fields
	if dto.LastSuccess != nil {
		stats.LastSuccess = JobStatusFromDTO(dto.LastSuccess)
	}
	if dto.LastFailure != nil {
		stats.LastFailure = JobStatusFromDTO(dto.LastFailure)
	}

	return stats
}

// JobStatusFromDTO converts an OpenAPI TgvalidatordJobStatus to a domain JobStatus.
func JobStatusFromDTO(dto *openapi.TgvalidatordJobStatus) *model.JobStatus {
	if dto == nil {
		return nil
	}

	status := &model.JobStatus{
		ID:      safeString(dto.Id),
		Message: safeString(dto.Message),
		Status:  safeString(dto.Status),
	}

	// Convert time fields
	if dto.StartedAt != nil {
		status.StartedAt = *dto.StartedAt
	}
	if dto.UpdatedAt != nil {
		status.UpdatedAt = *dto.UpdatedAt
	}
	if dto.TimeoutAt != nil {
		status.TimeoutAt = *dto.TimeoutAt
	}

	return status
}
